### ISSUE TITLE
Integrating Checkpoint Save Command into CLI

### ISSUE THREAD

**Issue by dbarer** [uchicago-cs/chiventure#153]
For this task, we want to get the save and load commands to work into CLI. The main thing from CLI's perspective is to understand what information checkpointing needs from user input to call their save and load function. For example, a file name, file type, etc. For the purposes of this sprint, there is no auto-save feature and save and load is only used if the user calls for it.

**Comment by jarvislam1999** [comment]
At the moment while we have a good idea of what our save functions would look like, our load function may change a lot from our last meeting with the instructor. 

We may have to create a new issue and take care of that first before coming over to here. You guys can certainly take care of our save function first, however.

**Comment by jarvislam1999** [comment]
Should I maybe add WDL into this since the loading part for checkpointing will involve both WDL and checkpointing from our last meeting? I think we will have to load the WDL first before calling our load function

**Comment by dbarer** [comment]
Update: Changed this issue to focus on implementing save, as Checkpointing needs to update their Load function. Will create an issue for load once it is finalized. However, Save is ready to be implemented. We discussed, and the only thing that we need to validate on the CLI side is that the file needs to be of type .dat, so we just need to make sure the filename string ends with that.

**Comment by jarvislam1999** [comment]
Some updates regarding the conversations from last week:
- Checkpointing talked to cli and agreed upon a course of action: checkpointing will try to finish and hand over the save function to CLI first by merging to master after the tests are passed, then do the same for load.
- CLI needs to be able to talk to action management so that whenever users type in save, it will invoke a function from WDL to load a WDL file into a game, and call our load function which will take that game and fill in our additional information. 
- We agreed on a file format such that every game would be written into a .dat file. While we are aware that this is not necessary since it's all binary, we still think this will help simplify matter between all the teams.

**Comment by jarvislam1999** [comment]
We have finished writing the save function, which can be easily accessed at checkpointing/include/save.h
But overall, the one function your team will use would be:

/*                                                                              
 * Save game into file                             
 *                                                                              
 * Parameters:                                                                  
 *   - g: pointer to game_t struct containing the game we want to save          
 *   - filename: pointer to string name of the save file                        
 *                                                                              
 * Return:                                                                      
 *   - 0 if successful                                                          
 *   - -1 if unsuccessful                                                       
 */
int save(game_t *game, char *filename);

Please let us know if this is good enough an interface and if you want to have anything else passed in. Once we agree, this issue should be closed. Please note that we haven't compiled or test our save code so it may not be functional, but I think since the interface will not change, it's save for you to take this function and write codes on it. Let us know if we should talk to action management about this instead.

**Comment by dbarer** [comment]
Update: Under the cli/dbarer/save branch I have added save to our supported command operations. This save process includes a validation step to make sure that the save filename ends with .dat. The logic makes sense to me, but need to get an updated save function on master in order to know 100% that everything works. But it seems like this will be done really soon, and we can move on to implementing load, which will be a little bit more complicated.

**Comment by jarvislam1999** [comment]
Update: load function should be done soon. However, the tests are not done yet. Will it be good enough for the save function to just be compilable but not tested to be pushed to master for now? @dbarer @mpcsh 

It would be a while until we get save to actually work. The people assigned to Makefile have not been working on getting code compilable yet.

**Comment by dbarer** [comment]
Update: I have merged the cli integration of save into the cli/dev branch. It is currently untested, since I did not have a working version of save, but it doesn't crash our building of the CLI. I will probably close this issue, as it is done from the cli side. We may need to make changes once a complete save function is given to us. I will then create an issue for loading and begin working on that.

@jarvislam1999 It doesn't really matter on my end whether a compilable save function is pushed to master or not, as far as I can tell, the work is done on the cli side, but it's not until the save function is tested that i'll know for sure.

**Comment by jarvislam1999** [comment]
In responding to your earlier comment, I think it is a good time for us to close this issue since you have received the interface for the function and that the CLI building process is not crashed. Once our code is all compiled and pushed to master, then save would probably automatically be compilable and hopefully, it will pass the tests as well.

Future steps include:
- Update each other via Slack once the save functions are good to go.
- Integrate load into CLI

**Comment by mpcsh** [comment]
Issue Score: ✓+

Comments:
Great work!
